### PR TITLE
Numerical stability Probit node and extension update rules Bernoulli node

### PR DESCRIPTION
**Pull request by bartvanerp**
This pull request implements the following:
- numerically stable procedure for calculating the messages and marginals around the Probit node.
- additional trivial update rules for the Bernoulli node.

The Probit node performs moment-matching for calculating the message. In specific situation (e.g. p=0, gamma >> 0) the normalization constant becomes 0 and extracting the moments is no longer feasible. The edits transform all operations to the log-domain for improved robustness.

For the tests to pass, two edits have been made in other parts of the code.
1. An `eltype()` function has been created for the PointMass distribution.
2. The original `eltype()` function in Distributions.jl has been overwritten for the Bernoulli distribution.

### CODE REVIEW

**Comment by bvdmitri** [comment]
We should not overwrite predefined methods for object from other libraries, we need some other mechanism here to promote_types or rewrite code in a different way as well without explicit `T`. I can work on that. Also, from `eltype` documentation:
```
Determine the type of the elements generated by iterating a collection of the given type.
```
I would say `eltype(PointMass([ 0.0, 0.0 ]))` should `Vector{Float64}`, not `Float64`

File: src/rules/probit/in.jl [hunk 1]
```diff
@@ -19,17 +20,23 @@ end
...
+        log_mom0_pz = logsumexp([log(1-p), log(2*p-1) + normlogccdf(-γ)])
+    elseif γ <= 0 && p > 0.5
+        log_mom0_pz = logsumexp([log(1-p), log(2*p-1) + normlogcdf(γ)])
+    elseif γ > 0 && p <= 0.5
+        log_mom0_pz = logsumexp([log(1-p) + normlogcdf(-γ), log(p) + normlogcdf(γ)])
+    else
+        log_mom0_pz = logsumexp([log(1-p) + normlogccdf(γ), log(p) + normlogcdf(γ)])
+    end
+    mom1_pz = mz + (2*p-1)*exp(log(vz) + normlogpdf(γ) - 0.5*log(1+vz) - log_mom0_pz)
+    mom2_pz = vz + mz^2 + (2*p-1)*2*mz*exp(log(vz) + normlogpdf(γ) - 0.5*log(1+vz) - log_mom0_pz) - (2p-1)*γ*exp(2*log(vz) + normlogpdf(γ) - log(1 + vz) - log_mom0_pz)
```
Comment (bvdmitri) [inline src/rules/probit/in.jl:39]: We can precalculate: `log(vz) + normlogpdf(γ) - 0.5*log(1+vz) - log_mom0_pz` only once.

File: src/rules/probit/in.jl [hunk 2]
```diff
@@ -19,17 +20,23 @@ end
...
-    umom1_g = normcdf(γ)*mz + vz*normpdf(γ)/sqrt(1+vz)
-    umom2_g = 2*mz*umom1_g + (vz - mz^2)*normcdf(γ) - vz^2*γ*normpdf(γ)/(1+vz)
-
-    # calculate moments of posterior
-    mom0_pz = 1 - p + (2*p - 1)*normcdf(γ)
-    mom1_pz = 1/mom0_pz*((1-p)*mz + (2*p-1)*umom1_g)
-    mom2_pz = 1/mom0_pz*((1-p)*(vz + mz^2) + (2*p-1)*umom2_g)
+    # calculate moments of g
+    if γ > 0 && p > 0.5
+        log_mom0_pz = logsumexp([log(1-p), log(2*p-1) + normlogccdf(-γ)])
```
Comment (bvdmitri) [inline src/rules/probit/in.jl:30]: I would create allocation-free version of logsumexp if possible. Now it allocates 2-element array every time

File: src/rules/probit/in.jl (hunk 1 above)
Comment (bartvanerp) [reply src/rules/probit/in.jl]: Good catch, commit (9302457) resolves this.

**Review by albertpod — COMMENTED** [review]
Thanks @bvdmitri for the review. 
Given new changes from @bartvanerp, PR seems good to me. Ready to merge.

File: src/rules/probit/in.jl (hunk 2 above)
Comment (bartvanerp) [reply src/rules/probit/in.jl:30]: Would be nice indeed. However, rewriting this function for a variable number of arguments actually increases the number of allocations and slows down performance significantly (~8x slower). If we really want to do this we could create separate functions for different number of arguments, but the `maximum` function requires a tuple/ an array anyway.

Commit (6d8bd11) implements a custom `logsumexp` function that is about 10% faster than the function from `StatsFuns`. Commits (222ac5b) and (0c4a9433) replace the original implementations with this custom `logsumexp` function

**Comment by bvdmitri** [comment]
Thanks @bartvanerp! I changed some parts of the original PR, imo its ready to merge!

Regarding my changes:
- I switched back to `logsumexp` from `StatsBase` and changed rule code to use tuple-based version instead.
- I changed `Float64` literals in rules to `Int64` literals to avoid unnecessary type promotions.
- I removed eltype for Bernoulli, changed eltype for PointMass - it returns an actual type of a point mass now
- minor stuff, better error message for typeof conversions in tests

**Review by bvdmitri — APPROVED** [review]

**Pull request merged**